src/renderer/generic.zig

```diff
@@ -2530,7 +2530,7 @@ pub fn Renderer(comptime GraphicsAPI: type) type {
 
                         // Advance our index until we reach or pass
                         // our current x position in the shaper cells.
-                        while (shaper_cells.?[shaper_cells_i].x < x) {
+                        while (shaper_cells.?[shaper_cells_i].x < x - run.offset) {
```

Comment:
This should be rephrased as `<...>.x + run.offset < x`, to avoid an underflow if x is ever less than run.offset.